File: src/main/test/com/edgelab/hospital/HospitalTests.java
package com.edgelab.hospital;

import com.edgelab.hospital.entities.Drug;
import com.edgelab.hospital.entities.Hospital;
import com.edgelab.hospital.entities.Patient;
import com.edgelab.hospital.entities.State;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.when;

public class HospitalTests {

    @Test
    void checkHospitalReportGeneration() {
        Patient patient1 = Mockito.mock(Patient.class);
        Patient patient2 = Mockito.mock(Patient.class);
        Patient patient3 = Mockito.mock(Patient.class);

        when(patient1.getState()).thenReturn(State.FEVER);
        when(patient2.getState()).thenReturn(State.HEALTHY);
        when(patient3.getState()).thenReturn(State.TUBERCULOSIS);

        Hospital hospital = new Hospital(Arrays.asList(patient1, patient2, patient3));
        String results = hospital.runSimulation(Arrays.asList(Drug.ANTIBIOTIC, Drug.ASPIRIN));
        assertEquals("F:1,H:1,D:0,T:1,X:0", results);
    }

    @Test
    void multiSimulations() {
        List<Patient> patients = Arrays.asList(new Patient(State.FEVER), new Patient(State.DIABETES), new Patient(State.TUBERCULOSIS));
        Hospital hospital = new Hospital(patients);

        String results1 = hospital.runSimulation(Arrays.asList(Drug.PARACETAMOL, Drug.INSULIN));
        assertEquals("F:0,H:1,D:1,T:1,X:0", results1);

        String results2 = hospital.runSimulation(Arrays.asList(Drug.ANTIBIOTIC, Drug.ASPIRIN));
        assertEquals("F:0,H:2,D:0,T:0,X:1", results2);
    }

}

File: src/test/kotlin/com/edgelab/hospital/HospitalTests.kt
package com.edgelab.hospital

import com.edgelab.hospital.Drug.*
import com.edgelab.hospital.State.*
import io.mockk.every
import io.mockk.spyk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class HospitalTests {

  @Test
  fun checkHospitalReportGeneration() {
    val patient1 = spyk(Patient(HEALTHY))
    val patient2 = spyk(Patient(HEALTHY))
    val patient3 = spyk(Patient(HEALTHY))

    every { patient1.state } returns FEVER
    every { patient2.state } returns HEALTHY
    every { patient3.state } returns TUBERCULOSIS

    val results = Hospital(listOf(ANTIBIOTIC, ASPIRIN)).runSimulation(listOf(patient1, patient2, patient3))
    assertThat(results.map { it.state }).containsExactlyInAnyOrderElementsOf(listOf(FEVER, HEALTHY, TUBERCULOSIS))
  }

  @Test
  fun multiSimulations() {
    val patients = listOf(Patient(FEVER), Patient(DIABETES), Patient(TUBERCULOSIS))
    assertThat(Hospital(listOf(PARACETAMOL, INSULIN)).runSimulation(patients).map { it.state })
      .containsExactlyInAnyOrderElementsOf(listOf(HEALTHY, DIABETES, TUBERCULOSIS))
    assertThat(Hospital(listOf(ANTIBIOTIC, ASPIRIN)).runSimulation(patients).map { it.state })
      .containsExactlyInAnyOrderElementsOf(listOf(HEALTHY, HEALTHY, DEAD))
  }

}
